core/src/main/scala/Diet.scala

```diff
@@ -0,0 +1,321 @@
+    @tailrec def genRange(a: A, b: A, xs: List[A])(implicit discrete: Discrete[A]): List[A] = {
+      if (discrete.compare(a, b) == EQ) {
+        xs ::: Nel(a, El())
+      } else if (discrete.adj(a, b)) {
+        xs ::: Nel(a, Nel(b, El()))
+      }
+      else {
+        genRange(discrete.succ(a), b, xs ::: (Nel(a, El())))
+      }
+    }
+
+    genRange(start, end, El())
+  }
+
+  override def reverse()(implicit discrete: Discrete[A]): List[A] = {
+    @tailrec def genRev(a: A, b: A, xs: List[A])(implicit discrete: Discrete[A]): List[A] = {
+      if (discrete.compare(a, b) == EQ) {
+        xs ::: Nel(a, El())
+      } else if (discrete.adj(a, b)) {
+        xs ::: Nel(a, Nel(b, El()))
+      }
+      else {
+        genRev(discrete.pred(a), b, xs ::: (Nel(a, El())))
+      }
+    }
+
+    genRev(end, start, El())
+  }
+
+  override def apply(start: A, end: A): DRange[A] = DRange.apply(start, end)
+}
+
+object DRange {
+  def apply[A](x: A, y: A) = new DRange[A](x, y)
+}
+
+/**
+ * Discrete Interval Encoding Tree (Diet).
+ * It stores subsets of types having a total order, a predecessor and a successor function described by Discrete[A]
+ *
+ * Diet is a binary search tree where each node contains a range of values and the set of all nodes is a set of
+ * disjoint sets.
+ *
+ * In the best case, when there are no "holes" in the stored set, the interval representation consists of just one
+ * single interval (node) and finding, inserting and deleting operations are O(1). In the worse case, where there
+ * are not two adjacent elements in the set, the representation is equivalent to a binary search tree.
+ */
+sealed abstract class Diet[A] {
+  import Diet._
+
+  val isEmpty: Boolean
+
+  /**
+   * verify x is a value is in the tree
+   */
+  def contains(x: A)(implicit discrete: Discrete[A]): Boolean = this match {
+    case EmptyDiet()          =>  false
+    case DietNode(a,b, l, r)  =>  (discrete.compare(x, a), discrete.compare(x, b)) match {
+      case (EQ, _)  =>  true
+      case (_, EQ)  =>  true
+      case (GT, LT) =>  true
+      case (LT, _)  =>  l.contains(x)
+      case (_, GT)  =>  r.contains(x)
+    }
+  }
+
+  /**
+   * return a list of all disjoint sets in the tree where each set is represented by ARange
+   */
+  def disjointSets(): List[DRange[A]] = this match {
+    case EmptyDiet()          =>  El()
+    case DietNode(x, y, l, r) =>  l.disjointSets() ::: (DRange(x, y) :: r.disjointSets())
+  }
+
+  /**
+   * convert tree in a sorted list from all disjoint sets in the tree
+   */
+  def toList()(implicit discrete: Discrete[A]): List[A] =
+    disjointSets().flatMap(lst => lst.generate())
+
+  /**
+   * add new value range [x, y] to de tree. If x > y then it will add range [y, x]
+   */
+  def add(x: A, y: A)(implicit discrete: Discrete[A]): Diet[A] = {
+    if (discrete.compare(x, y) == EQ) {
+      add(x)
+    }
+    else if (discrete.compare(x, y) == LT) {
+      add(x).add(discrete.succ(x), y)
```

Comment:
The original paper doesn't talk about adding range, this addRange is very naive, it needs improvements, but not sure how to do it yet. We can remove it if you want to.